File: Valid Perfect Square - Leetcode 367/Valid Perfect Square - Leetcode 367.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        l = 1
        r = num

        while l <= r:
            m = (l+r) // 2
            m_squared = m * m

            if num == m_squared:
                return True
            elif m_squared < num:
                l = m + 1
            else:
                r = m - 1
        
        return False # Time: O(log n), Space: O(1)

File: Valid Perfect Square - Leetcode 367/Valid Perfect Square - Leetcode 367.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        long l = 1; // Use long to prevent overflow
        long r = num;

        while (l <= r) {
            long m = (l + r) / 2;
            long mSquared = m * m;

            if (num == mSquared) {
                return true;
            } else if (mSquared < num) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }

        return false;
    }
};
